### PR TITLE
feat(Cache): better error handling

### PR DESCRIPTION
**Pull request by NathanWalker**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [x] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/DevelopmentWorkflow.md#running-unit-tests.
- [ ] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/WritingUnitTests.md.

## What is the current behavior?

Usage of the image `Cache` class can result in many app problems due to invalid error handling conditions.

## What is the new behavior?

Proper error handling can be wired up to circumvent troubling app conditions, ie:

```
this._cache.push((<DownloadRequest> {
  key: url,
  url: url,
  completed: (image: any, key: string) => {                                                       
    if (url === key) {
       // handle success
    }
  },
  error: (key: string) => {
    if (url === key) {
      // we can now properly handle error for failed url request here
    }
  }
}));
```

These changes are dependent on widgets PR here: https://github.com/NativeScript/tns-core-modules-widgets/pull/147

### CODE REVIEW

**Comment by NathanWalker** [comment]
This helps avoid these app crashes as well:
```
System.err: StackTrace:
System.err: Frame: function:'push.../node_modules/tns-core-modules/ui/image-cache/image-cache.js.Cache.set', file:'file:///data/data/org.nativescript.app/files/app/vendor.js', line: 134088, column: 21
```

**Comment by vakrilov** [comment]
Hi and thanks for the PR!

The team have been quite busy baking the upcoming NativeScript 5.0 release. We hope you love it as much as we do! The major release is the reason we taking longer with PR reviews. We will resume active reviewing in the following weeks. 

Thanks for your patience and for being awesome contributor!

**Comment by vchimev** [comment]
Hey @NathanWalker,

Let's get merged the related [pull request](https://github.com/NativeScript/tns-core-modules-widgets/pull/147) in `tns-core-modules-widgets` to facilitate the work here.

Could you also share how to reproduce the errors you experience?

Thank you!

File: tns-core-modules/ui/image-cache/image-cache.android.ts [hunk 1]
```diff
@@ -61,7 +72,15 @@ export class Cache extends common.Cache {
     }
 
     public set(key: string, image: any): void {
-        this._cache.put(key, image);
+      try {
+        if (key && key != null && image && image != null) {
```
Comment (vchimev) [inline tns-core-modules/ui/image-cache/image-cache.android.ts]: No need of this `if()` statement as `set()` is will be executed only when `key` and `image` are available [here](https://github.com/NativeScript/NativeScript/blob/842f9132541e080d72448cf7d518998a9cd2bf12/tns-core-modules/ui/image-cache/image-cache-common.ts#L144).

**Comment by vchimev** [comment]
Hey @NathanWalker,

I committed a few updates on the API and implementation that we think are required and would expect in future pull requests.

Could you please handle my comments above to proceed further?

Thank you!

**Comment by vchimev** [comment]
Hey @NathanWalker, is there any update here?

**Comment by NathanWalker** [comment]
@vchimev ok got this updated now. The changes are based on this PR which I completely agree with:
https://github.com/NativeScript/tns-core-modules-widgets/pull/149

Let me know if you need anything else here.

**Comment by vchimev** [comment]
test

**Comment by SvetoslavTsenov** [comment]
test

**Pull request merged**